### DISCUSSION TITLE
Can i extend the remote reactor?

### DISCUSSION THREAD

**Discussion by burakakca** [lf-lang/lingua-franca#2446] [Q&A]
For example;
 
```
reactor PrintMessageWithDetector(offset: time = 10 msec) extends PrintMessage at 192.168.88.32 {
```

**Comment by erlingrj** [comment]
Yes, the network address of a federate does not affect the ability to extend other reactors. However, your syntax is not entirely correct. To specify the host name of a federate you use `at 192.168.88.32` on the federate *instance*, not the reactor *declaration*. There is an exception, you can do `federated reactor MyFederation at 192.168.88.33`, but this is a way of specifying the hostname of the RTI, not of any of the federates. In your case you would want to do:

```
// Reactor declaration, can extend any other reactor, no details about the hostname/IP address give here.
reactor PrintMessageWithDetector(offset: time = 10 msec) extends PrintMessage {
   // Add your reactor elements here
}

federated reactor {
  // Federate instance, here you specify the hostname or IP address where the federate will run.
  printer = new PrintMessageWithDetector() at 192.168.88.33
}
```

For more info, check out or docs: https://www.lf-lang.org/docs/writing-reactors/distributed-execution#specifying-rti-hosts

Hope this helps,

**Comment by burakakca** [comment]
Let me explain again to understand if we are on the same page, there is a reactor that will work on the remote server and I am trying to extend it, so for example, "PrintMessage" is on the remote server and I am extending it and using it as "PrintMessageWithDetector".

What you are describing is an extended reactor that will work on the remote server

**Reply by erlingrj** [reply]
I am not 100% sure I understand what you mean here. But it is important to point the difference between a reactor *declaration* and a reactor *instance*. A reactor declaration is like a class in C++ and a reactor instance is like an instance of that class.  

In your code, you specify the IP address on a reactor declaration using `at`, which does not work in Lingua Franca and should result in a compiler error. Instead, you should specify the IP address when you create an instance of `PrintMessageWithDetector`. In this way, you can create many instances of `PrintMessageWithDetector` and put them on different servers. So to your question, when you extend a reactor declaration, there is no notion of an IP address, so you cannot inherit any IP address either. It is only when you create instances of reactors that you can assign IP addresses. If you want to create two federates,`PrintMessage` and `PrintMessageWithDetector`, and put both on the same remote server you must create two separate reactor instances and specify the same IP address for them.

In the code snippet I shared I show you how you specify the IP address on the reactor instance.

**Reply by burakakca** [reply]
I understand that how to use remote reactors. My first post is an example. I know that is not gonna work.

I have another diffrent question;
- Can we use diffrent targets in the same lf file with importing. ( i have federated code  in c++ and i'm imorting ts and py lf files into that c++ lf file)
- I couldn't get it to work as much as I tried. Will it be possible in the future? or not?

**Reply by erlingrj** [reply]
Currently, there is no support for LF programs using several target languages (federated or not). This is something we would like to support in the future, but I am not aware of a concrete plan for it yet.

**Reply by edwardalee** [reply]
There is a rather stale `mixed-federation` branch that has a starting point for this (little more than a concept demonstration). It would work with a top-level target called `mixed`, a federated reactor that specifies on each connection the serialization strategy (such as protobufs), and imported LF files that can be each implemented in a different target language, as long as it supports the same serialization strategy.  There is a fair amount of coding to make all this work, and we could use some help.

**Reply by burakakca** [reply]
That's great thanks. @edwardalee
